sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/read/SnowflakeIOReadTest.java

```diff
@@ -0,0 +1,268 @@
+        SnowflakeIO.<GenericRecord>read(snowflakeService)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .fromTable(FAKE_TABLE)
+            .withStagingBucketName(stagingBucketName)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper()));
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testConfigIsMissingFromTableOrFromQuery() {
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage("fromTable() or fromQuery() is required");
+
+    pipeline.apply(
+        SnowflakeIO.<GenericRecord>read(snowflakeService)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .withStagingBucketName(stagingBucketName)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema())));
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testConfigIsMissingDataSourceConfiguration() {
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage(
+        "withDataSourceConfiguration() or withDataSourceProviderFn() is required");
+
+    pipeline.apply(
+        SnowflakeIO.<GenericRecord>read(snowflakeService)
+            .fromTable(FAKE_TABLE)
+            .withStagingBucketName(stagingBucketName)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema())));
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testConfigContainsFromQueryAndFromTable() {
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage("fromTable() and fromQuery() are not allowed together");
+
+    pipeline.apply(
+        SnowflakeIO.<GenericRecord>read(snowflakeService)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .fromQuery("")
+            .fromTable(FAKE_TABLE)
+            .withStagingBucketName(stagingBucketName)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema())));
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testTableDoesntExist() {
+    exceptionRule.expect(PipelineExecutionException.class);
+    exceptionRule.expectMessage("SQL compilation error: Table does not exist");
+
+    pipeline.apply(
+        SnowflakeIO.<GenericRecord>read(snowflakeService)
+            .withDataSourceConfiguration(dataSourceConfiguration)
+            .fromTable("NON_EXIST")
+            .withStagingBucketName(stagingBucketName)
+            .withIntegrationName(integrationName)
+            .withCsvMapper(getCsvMapper())
+            .withCoder(AvroCoder.of(AvroGeneratedUser.getClassSchema())));
+
+    pipeline.run();
+  }
+
+  @Test
+  public void testReadWithConfigIsProper() {
+    PCollection<GenericRecord> items =
```

Comment:
Please make sure that all read paths are covered by tests.

Reply:
I've just added tests for `withQuery` as it was missing and fixed tests in general